File: Freedom/微信/我/Setting/WXBgSettingViewController.m
//  FreedomBackgroundSettingViewController.m
//  Freedom
//  Created by Super on 16/3/19.
#import "WXBgSettingViewController.h"
#import <BlocksKit/BlocksKit+UIKit.h>
#import "WXChatViewController.h"
#import "WXModes.h"
#import <ReactiveCocoa/ReactiveCocoa-Swift.h>
#import "WXCommonSettingViewController.h"
#import "WXBaseViewController.h"
#import "NSFileManager+expanded.h"
#define     SPACE_EDGE                      10
#define     WIDTH_COLLECTIONVIEW_CELL       (APPW - SPACE_EDGE * 2) / 3 * 0.98
#define     SPACE_COLLECTIONVIEW_CELL       (APPW - SPACE_EDGE * 2 - WIDTH_COLLECTIONVIEW_CELL * 3) / 2
@interface WXChatBackgroundSelectViewController : WXBaseViewController<UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout>
- (void)registerCellForCollectionView:(UICollectionView *)collectionView;
@property (nonatomic, strong) NSMutableArray *data;
@property (nonatomic, strong) UICollectionView *collectionView;
@end
@interface WXChatBackgroundSelectViewController ()
@end
@implementation WXChatBackgroundSelectViewController
- (void)viewDidLoad{
    [super viewDidLoad];
    [self.navigationItem setTitle:@"选择背景图"];
    [self.collectionView setBackgroundColor:RGBACOLOR(46.0, 49.0, 50.0, 1.0)];
    [self.view addSubview:self.collectionView];
    
    [self p_addMasonry];
}
#pragma mark - 
- (void)p_addMasonry{
    [self.collectionView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.mas_equalTo(self.view);
    }];
}
#pragma mark - 
- (UICollectionView *)collectionView{
    if (_collectionView == nil) {
        UICollectionViewFlowLayout *layout = [[UICollectionViewFlowLayout alloc] init];
        [layout setSectionHeadersPinToVisibleBounds:YES];
        [layout setItemSize:CGSizeMake(WIDTH_COLLECTIONVIEW_CELL, WIDTH_COLLECTIONVIEW_CELL)];
        [layout setMinimumInteritemSpacing:SPACE_COLLECTIONVIEW_CELL];
        [layout setMinimumLineSpacing:SPACE_COLLECTIONVIEW_CELL];
        [layout setSectionInset:UIEdgeInsetsMake(0, SPACE_EDGE, 0, SPACE_EDGE)];
        _collectionView = [[UICollectionView alloc] initWithFrame:CGRectZero collectionViewLayout:layout];
        [_collectionView setBackgroundColor:[UIColor clearColor]];
        [_collectionView setDataSource:self];
        [_collectionView setDelegate:self];
        [_collectionView setShowsHorizontalScrollIndicator:NO];
        [_collectionView setShowsHorizontalScrollIndicator:NO];
        [_collectionView setAlwaysBounceVertical:YES];
    }
    return _collectionView;
}
- (void)registerCellForCollectionView:(UICollectionView *)collectionView{
    [collectionView registerClass:[UICollectionViewCell class] forCellWithReuseIdentifier:@"TLChatBackgroundSelectCell"];
}
#pragma mark - 
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section{
    return self.data.count;
}
- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath{
    UICollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"TLChatBackgroundSelectCell" forIndexPath:indexPath];
    return cell;
}
@end
@interface WXBgSettingViewController () <UINavigationControllerDelegate,UIImagePickerControllerDelegate>
@end
@implementation WXBgSettingViewController
- (void)viewDidLoad {
    [super viewDidLoad];
    [self.navigationItem setTitle:@"聊天背景"];
    
    self.data = [WXCommonSettingHelper chatBackgroundSettingData];
}
#pragma mark - Delegate
//MARK: UITableViewDataSource
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    if (self.data.count > 0) {
        return self.partnerID.length > 0 ? self.data.count - 1 : self.data.count;
    }
    return 0;
}
- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingImage:(UIImage *)image editingInfo:(nullable NSDictionary<NSString *,id> *)editingInfo{
    [picker dismissViewControllerAnimated:YES completion:^{
        UIImage *image = [editingInfo objectForKey:UIImagePickerControllerOriginalImage];
        [self p_setChatBackgroundImage:image];
    }];
}
- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary<NSString *,id> *)info{
    [picker dismissViewControllerAnimated:YES completion:^{
        UIImage *image = [info objectForKey:UIImagePickerControllerOriginalImage];
        [self p_setChatBackgroundImage:image];
    }];
}
- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker{
    [picker dismissViewControllerAnimated:YES completion:nil];
}
//MARK: UITableViewDelegate
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    WXSettingItem *item = [self.data[indexPath.section] objectAtIndex:indexPath.row];
    if ([item.title isEqualToString:@"选择背景图"]) {
        WXChatBackgroundSelectViewController *bgSelectVC = [[WXChatBackgroundSelectViewController alloc] init];
        [self setHidesBottomBarWhenPushed:YES];
        [self.navigationController pushViewController:bgSelectVC animated:YES];
    }else if ([item.title isEqualToString:@"从手机相册中选择"]) {
        UIImagePickerController *imagePickerController = [[UIImagePickerController alloc] init];
        [imagePickerController setSourceType:UIImagePickerControllerSourceTypePhotoLibrary];
        imagePickerController.delegate = self;
        [self presentViewController:imagePickerController animated:YES completion:nil];
    }else if ([item.title isEqualToString:@"拍一张"]) {
        UIImagePickerController *imagePickerController = [[UIImagePickerController alloc] init];
        if(![UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypeCamera]) {
            [SVProgressHUD showErrorWithStatus:@"相机初始化失败"];
        }else{
            [imagePickerController setSourceType:UIImagePickerControllerSourceTypeCamera];
            imagePickerController.delegate = self;
            [self presentViewController:imagePickerController animated:YES completion:nil];
        }
    }else if ([item.title isEqualToString:@"将背景应用到所有聊天场景"]) {
        [self showAlerWithtitle:nil message:nil style:UIAlertControllerStyleActionSheet ac1:^UIAlertAction *{
            return [UIAlertAction actionWithTitle:@"将背景应用到所有聊天场景" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                for (NSString *key in [NSUserDefaults standardUserDefaults].dictionaryRepresentation.allKeys) {
                    if ([key hasPrefix:@"CHAT_BG_"] && ![key isEqualToString:@"CHAT_BG_ALL"]) {
                        [[NSUserDefaults standardUserDefaults] removeObjectForKey:key];
                    }
                }
                [[WXChatViewController sharedChatVC] resetChatVC];
            }];
        } ac2:^UIAlertAction *{
            return [UIAlertAction actionWithTitle:@"取消" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {

            }];
        } ac3:nil completion:nil];
    }
    [tableView deselectRowAtIndexPath:indexPath animated:NO];
}
#pragma mark - Private Methods -
- (void)p_setChatBackgroundImage:(UIImage *)image{
    
    NSData *imageData = (UIImagePNGRepresentation(image) ? UIImagePNGRepresentation(image) :UIImageJPEGRepresentation(image, 1));
    NSString *imageName = [NSString stringWithFormat:@"%lf.jpg", [NSDate date].timeIntervalSince1970];
    NSString *imagePath = [NSFileManager pathUserChatBackgroundImage:imageName];;
    [[NSFileManager defaultManager] createFileAtPath:imagePath contents:imageData attributes:nil];
    
    //TODO: 临时写法
    if (self.partnerID.length > 0) {
        [[NSUserDefaults standardUserDefaults] setObject:imageName forKey:[@"CHAT_BG_" stringByAppendingString:self.partnerID]];
    }else{
        [[NSUserDefaults standardUserDefaults] setObject:imageName forKey:@"CHAT_BG_ALL"];
    }
    
    [[WXChatViewController sharedChatVC] resetChatVC];
}
@end

File: Freedom/微信/我/Setting/WXBgSettingViewController.swift
//
//  WXBgSettingViewController.swift
//  Freedom
import SnapKit
import Foundation
import XExtension
class WXChatBackgroundSelectViewController: WXBaseViewController, UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout {
    var data: [AnyHashable] = []
    override func viewDidLoad() {
        super.viewDidLoad()
        navigationItem.title = "选择背景图"
        let layout = UICollectionViewFlowLayout()
        layout.sectionHeadersPinToVisibleBounds = true
        layout.itemSize = CGSize(width: (APPW - 10 * 2) / 3 * 0.98, height: (APPW - 10 * 2) / 3 * 0.98)
        layout.minimumInteritemSpacing = (APPW - 10 * 2 - (APPW - 10 * 2) / 3 * 0.98 * 3) / 2
        layout.minimumLineSpacing = (APPW - 10 * 2 - (APPW - 10 * 2) / 3 * 0.98 * 3) / 2
        layout.sectionInset = UIEdgeInsets(top: 0, left: 10, bottom: 0, right: 10)
        collectionView = BaseCollectionView(frame: CGRect.zero, collectionViewLayout: layout)
        collectionView.backgroundColor = UIColor.clear
        collectionView.dataSource = self
        collectionView.delegate = self
        collectionView.showsHorizontalScrollIndicator = false
        collectionView.showsHorizontalScrollIndicator = false
        collectionView.alwaysBounceVertical = true
        collectionView.backgroundColor = UIColor(46.0, 49.0, 50.0, 1.0)
        view.addSubview(collectionView)
        collectionView.snp.makeConstraints { (make) in
            make.edges.equalTo(self.view)
        }
    }
    func registerCell(for collectionView: UICollectionView) {
        collectionView.register(UICollectionViewCell.self, forCellWithReuseIdentifier: "TLChatBackgroundSelectCell")
    }
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return data.count
    }
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell: UICollectionViewCell = collectionView.dequeueReusableCell(withReuseIdentifier: "TLChatBackgroundSelectCell", for: indexPath)
        return cell
    }
}
class WXBgSettingViewController: WXSettingViewController, UINavigationControllerDelegate, UIImagePickerControllerDelegate {
    //若为nil则全局设置，否则只给对应好友设置
    var partnerID = ""
    override func viewDidLoad() {
        super.viewDidLoad()
        navigationItem.title = "聊天背景"
        data = WXCommonSettingHelper.chatBackgroundSettingData()
    }
    override func numberOfSections(in tableView: UITableView) -> Int {
        if data.count > 0 {
            return partnerID.count > 0 ? data.count - 1 : data.count
        }
        return 0
    }

    func imagePickerController(_ picker: UIImagePickerController, didFinishPicking image: UIImage, editingInfo: [String : Any]) {
        picker.dismiss(animated: true) {
            self.p_setChatBackgroundImage(image)
        }
    }

    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
        picker.dismiss(animated: true) {
            let image = info[.originalImage] as! UIImage
            self.p_setChatBackgroundImage(image)
        }
    }
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let item = data[indexPath.section].items[UInt(indexPath.row)]
        if (item.title == "选择背景图") {
            let bgSelectVC = WXChatBackgroundSelectViewController()
            hidesBottomBarWhenPushed = true
            navigationController?.pushViewController(bgSelectVC, animated: true)
        } else if (item.title == "从手机相册中选择") {
            let imagePickerController = UIImagePickerController()
            imagePickerController.sourceType = .photoLibrary
            imagePickerController.delegate = self
            present(imagePickerController, animated: true)
        } else if (item.title == "拍一张") {
            let imagePickerController = UIImagePickerController()
            if !UIImagePickerController.isSourceTypeAvailable(.camera) {
                noticeError("相机初始化失败")
            } else {
                imagePickerController.sourceType = .camera
                imagePickerController.delegate = self
                present(imagePickerController, animated: true)
            }
        } else if (item.title == "将背景应用到所有聊天场景") {
            let alert = UIAlertController("", "", T1: "将背景应用到所有聊天场景", T2: "取消", confirm1: {
                for key: String in UserDefaults.standard.dictionaryRepresentation().keys {
                    if key.hasPrefix("CHAT_BG_") && !(key == "CHAT_BG_ALL") {
                        UserDefaults.standard.removeObject(forKey: key)
                    }
                }
                WXChatViewController.shared.resetChatVC()
            }, confirm2: {

            })
            self.present(alert, animated: true, completion: nil)
        }
        tableView.deselectRow(at: indexPath, animated: false)
    }
    func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
        picker.dismiss(animated: true)
    }
    func p_setChatBackgroundImage(_ image: UIImage) {
        var imageData: Data?
        imageData = image.pngData() != nil ? image.pngData() : image.jpegData(compressionQuality: 1)
        let imageName = String(format: "%lf.jpg", Date().timeIntervalSince1970)
        let imagePath = FileManager.pathUserChatBackgroundImage(imageName)
        FileManager.default.createFile(atPath: imagePath, contents: imageData, attributes: nil)
        if partnerID.count > 0 {
            UserDefaults.standard.set(imageName, forKey: "CHAT_BG_" + (partnerID))
        } else {
            UserDefaults.standard.set(imageName, forKey: "CHAT_BG_ALL")
        }
        WXChatViewController.shared.resetChatVC()
    }
}
